File: sdk/typescript/src/formatter.ts
import type { ApplyResult, ApplyRule } from './types.js';

/**
 * Apply template rules to credential values.
 * Template syntax: ${key} is replaced with values[key] (empty string if missing).
 */
export function applyRules(values: Record<string, string>, rules: ApplyRule[]): ApplyResult {
    const headers: Record<string, string> = {};
    let query: Record<string, string> | undefined;
    let body: Record<string, string> | undefined;

    for (const rule of rules) {
        const action = rule.action ?? 'set';
        const interpolated = interpolate(rule.value, values);

        if (rule.in === 'header') {
            if (action === 'remove') {
                delete headers[rule.name];
            } else if (action === 'append') {
                const existing = headers[rule.name];
                headers[rule.name] = existing ? `${existing}; ${interpolated}` : interpolated;
            } else {
                headers[rule.name] = interpolated;
            }
        } else if (rule.in === 'query') {
            if (!query) query = {};
            if (action === 'remove') {
                delete query[rule.name];
            } else if (action === 'append') {
                const existing = query[rule.name];
                query[rule.name] = existing ? `${existing}; ${interpolated}` : interpolated;
            } else {
                query[rule.name] = interpolated;
            }
        } else if (rule.in === 'body') {
            if (!body) body = {};
            if (action === 'remove') {
                delete body[rule.name];
            } else if (action === 'append') {
                const existing = body[rule.name];
                body[rule.name] = existing ? `${existing}; ${interpolated}` : interpolated;
            } else {
                body[rule.name] = interpolated;
            }
        }
    }

    return { headers, ...(query ? { query } : {}), ...(body ? { body } : {}) };
}

function interpolate(template: string, values: Record<string, string>): string {
    return template.replace(/\$\{([^}]+)\}/g, (_, key) => values[key] ?? '');
}

File: sdk/python/src/sigcli_sdk/formatter.py
import re
from typing import Optional
from .types import ApplyRule, ApplyResult

def apply_rules(values: dict[str, str], rules: list[ApplyRule]) -> ApplyResult:
    """Apply template rules to credential values.

    Template syntax: ${key} is replaced with values[key] (empty string if missing).
    """
    headers: dict[str, str] = {}
    query: Optional[dict[str, str]] = None
    body: Optional[dict[str, str]] = None

    for rule in rules:
        action = rule.action or "set"
        interpolated = _interpolate(rule.value, values)

        if rule.in_ == "header":
            if action == "remove":
                headers.pop(rule.name, None)
            elif action == "append":
                existing = headers.get(rule.name)
                headers[rule.name] = f"{existing}; {interpolated}" if existing else interpolated
            else:
                headers[rule.name] = interpolated
        elif rule.in_ == "query":
            if query is None:
                query = {}
            if action == "remove":
                query.pop(rule.name, None)
            elif action == "append":
                existing = query.get(rule.name)
                query[rule.name] = f"{existing}; {interpolated}" if existing else interpolated
            else:
                query[rule.name] = interpolated
        elif rule.in_ == "body":
            if body is None:
                body = {}
            if action == "remove":
                body.pop(rule.name, None)
            elif action == "append":
                existing = body.get(rule.name)
                body[rule.name] = f"{existing}; {interpolated}" if existing else interpolated
            else:
                body[rule.name] = interpolated

    return ApplyResult(headers=headers, query=query, body=body)

def _interpolate(template: str, values: dict[str, str]) -> str:
    return re.sub(r"\$\{([^}]+)\}", lambda m: values.get(m.group(1), ""), template)
